oprf/group/suite.go

```diff
@@ -0,0 +1,251 @@
+type HashToPoint interface {
+	Hash(msg []byte) (*Element, error)
+}
+
+type eccHasher struct {
+	suite  *Ciphersuite
+	hasher h2c.HashToPoint
+}
+
+func (h eccHasher) Hash(in []byte) (*Element, error) {
+	q := h.hasher.Hash(in)
+	x := q.X().Polynomial()
+	y := q.Y().Polynomial()
+
+	p := &Element{h.suite, new(big.Int), new(big.Int), true}
+	p.x.Set(x[0])
+	p.y.Set(y[0])
+
+	if !p.IsValid() {
+		return nil, errors.New("invalid point")
+	}
+
+	return p, nil
+}
+
+// HashToGroup performs a transformation to encode bytes as a Element object in the
+// group.
+func (c *Ciphersuite) HashToGroup(in []byte) (*Element, error) {
+	hasher, err := getH2CSuite(c)
+	if err != nil {
+		return nil, err
+	}
+
+	p, err := hasher.Hash(in)
+	if err != nil {
+		return nil, err
+	}
+
+	return p, nil
+}
+
+// HashToScalar performs a transformation to encode bytes as a Scalar object in the
+// appropriate group.
+func (c *Ciphersuite) HashToScalar([]byte) (*Scalar, error) {
+	return &Scalar{}, nil
+}
+
+// RandomScalar samples a random scalar value from the field of scalars defined by the
+// group order.
+// TODO: not constant time
+func (c *Ciphersuite) RandomScalar() *Scalar {
+	N := c.Order()
+	bitLen := N.x.BitLen()
+	byteLen := (bitLen + 7) >> 3
+	buf := make([]byte, byteLen)
+
+	// rejection sampling
+	for {
+		_, err := io.ReadFull(rand.Reader, buf)
+		if err != nil {
+			panic("scalar generation failed")
+		}
+		// Mask to account for field sizes that are not a whole number of bytes.
+		var mask = []byte{0xff, 0x1, 0x3, 0x7, 0xf, 0x1f, 0x3f, 0x7f}
+		buf[0] = buf[0] & mask[bitLen%8]
+
+		// Check if scalar is in the correct range.
+		if new(big.Int).SetBytes(buf).Cmp(N.x) >= 0 {
+			continue
+		}
+		break
+	}
+
+	x := new(big.Int).SetBytes(buf)
+	return &Scalar{c, x}
+}
+
+// ScalarMultBase performs a scalar multiplication of the Generator with some scalar
+// input
+func (c *Ciphersuite) ScalarMultBase(s *Scalar) *Element {
+	gen := c.Generator()
+
+	return gen.ScalarMult(s)
+}
+
+// ByteLength returns the ByteLength of objects associated with the Ciphersuite.
+func (c *Ciphersuite) ByteLength() int {
+	return (c.curve.Params().BitSize + 7) / 8
+}
+
+// FinalizeHash computes the final hash for the suite
+func FinalizeHash(c *Ciphersuite, data, iToken, info, ctx []byte) []byte {
+	h := c.hash
+	lenBuf := make([]byte, 2)
+
+	binary.BigEndian.PutUint16(lenBuf, uint16(len(data)))
+	_, _ = h.Write(lenBuf)
+	_, _ = h.Write(data)
+
+	binary.BigEndian.PutUint16(lenBuf, uint16(len(iToken)))
+	_, _ = h.Write(lenBuf)
+	_, _ = h.Write(iToken)
+
+	binary.BigEndian.PutUint16(lenBuf, uint16(len(info)))
+	_, _ = h.Write(lenBuf)
+	_, _ = h.Write(info)
+
+	dst := []byte("VOPRF05-Finalize-")
+	dst = append(dst, ctx...)
+
+	binary.BigEndian.PutUint16(lenBuf, uint16(len(dst)))
+	_, _ = h.Write(lenBuf)
+	_, _ = h.Write(dst)
+
+	return h.Sum(nil)
+}
+
+// NewSuite creates a new ciphersuite for the requested name.
+func NewSuite(name string, ctx []byte) (*Ciphersuite, error) {
```

Comment:
Why don't we create suites with `SuiteID` (uint16) values?

Reply:
Sure! I looks a little weird, but it works.